doc/design/eventdriven-callbacks.md

```diff
@@ -0,0 +1,171 @@
+      }
+
+      myCallbackReactal.attachEvent(subscriber, iox::popo::SubscriberEvent::HAS_SAMPLE_RECEIVED, sampleReceived);
+      ```
+ - concurrently: detach a callback at any time from anywhere. This means in particular
+    - Detaching a callback from within a callback.
+    - [robust] When the detach call returns we guarantee that the callback is never called
+      again. E.g. blocks till the callback is removed, if the callback is concurrently 
+      running it will block until the callback is finished and removed.
+      Exception: If a callback detaches itself it blocks until the callback is removed 
+                 and not until the callback is finished.
+    - Calling detach means that after that call the callback is no longer attached
+      even when it was not attached in the first place. Therefore it will always succeed.
+
+ - When the ReactAl goes out of scope it should detach itself from every class 
+     to which it was attached via a provided callback (like in the WaitSet).
+
+ - When the class which is attached to the ReactAl goes out of scope it should 
+    detach itself from the ReactAl via a provided callback (like in the WaitSet).
+
+### Code Example
+```cpp
+ReactAl myCallbackReactal;
+iox::popo::UntypedSubscriber mySubscriber;
+
+void sampleReceived(iox::popo::UntypedSubscriber & subscriber) {
+  subscriber.take().and_then([&](auto & sample){
+    std::cout << "received " << std::hex << sample->payload() << std::endl;
+  });
+}
+
+void waitForSubscribtion(iox::popo::UntypedSubscriber & subscriber) {
+  std::cout << "subscribed!\n";
+  myCallbackReactal.attachEvent(subscriber, iox::popo::SubscriberEvent::HAS_SAMPLE_RECEIVED, sampleReceived);
+  myCallbackReactal.attachEvent(subscriber, iox::popo::SubscriberEvent::UNSUBSCRIBED, waitForSubscribtion);
+  myCallbackReactal.detachEvent(subscriber, iox::popo::SubscriberEvent::SUBSCRIBED);
+}
+
+void waitForUnsubscribe(iox::popo::UntypedSubscriber & subscriber) {
+  std::cout << "unsubscribed from publisher\n";
+  myCallbackReactal.attachEvent(subscriber, iox::popo::SubscriberEvent::SUBSCRIBED, waitForSubscribtion);
+  myCallbackReactal.detachEvent(subscriber, iox::popo::SubscriberEvent::HAS_SAMPLE_RECEIVED);
+  myCallbackReactal.detachEvent(subscriber, iox::popo::SubscriberEvent::UNSUBSCRIBED);
+}
+
+int main() {
+  myCallbackReactal.attachEvent(mySubscriber, iox::popo::SubscriberEvent::SUBSCRIBED, waitForSubscribtion);
+}
+
+```
+
+## Overall changes
+### Condition Variable
+
+  - add member `std::atomic_bool m_triggeredBy[MAX_CALLBACKS];`
+
+    **Reason:**
+      - The reactal will iterate through this array to know who was triggering it 
+      - It is cache local so very fast to iterate.
+      - WaitSet approach are callbacks which have overhead implementation 
+        and performance wise.
+
+    **Alternative:**
+      - Create a new class or provide a template bool array size argument to the 
+        ConditionVariable to be more flexible and memory efficient.
```

Comment:
Why not extending the `ConditionVariable` classes like we are doing it with the chunk building blocks. I'm not happy with adapting the condition variable API. We should try to follow the STL and extend the STL implementation e.g. `NotifySpecific(id)`. IMHO I would name this derived class `EventSignaler`.

Reply:
I added a reasoning for this approach.